src/ingest/src/main/scala/osmesa/functions/osm/package.scala

```diff
@@ -0,0 +1,302 @@
+        Set("yes", "true", "1").contains(tags("area").toLowerCase)
+      case _ =>
+        // see https://github.com/osmlab/id-area-keys (values are inverted)
+        val matchingKeys = tags.keySet.intersect(AREA_KEYS.keySet)
+        matchingKeys.exists(k => !AREA_KEYS(k).contains(tags(k)))
+    }
+
+  val isArea: UserDefinedFunction = udf(_isArea)
+
+  private val _isMultiPolygon = (tags: Map[String, String]) =>
+    tags.contains("type") && Set("boundary", "multipolygon").contains(tags("type").toLowerCase)
+
+  val isMultiPolygon: UserDefinedFunction = udf(_isMultiPolygon)
+
+  val buildWay: UserDefinedFunction = udf((nodes: Seq[Row], isArea: Boolean) => {
+    val geom = nodes
+      .sortWith(_.getAs[Int]("idx") < _.getAs[Int]("idx"))
+      .map(row =>
+        (Option(row.get(row.fieldIndex("lon"))).map(_.asInstanceOf[Double]).getOrElse(Double.NaN),
+          Option(row.get(row.fieldIndex("lat"))).map(_.asInstanceOf[Double]).getOrElse(Double.NaN)))
+    match {
+      // drop ways with invalid coordinates
+      // check for nulls (or add a coalesce to the select after join); since posexplode_outer is used, there will
+      // always be a row but it may be empty
+      case coords if coords.exists { case (x, y) => x.equals(Double.NaN) || y.equals(Double.NaN) } => None
+      // drop ways that don't contain valid geometries
+      case coords if coords.isEmpty => Some("LINESTRING EMPTY".parseWKT)
+      case coords if coords.length == 1 =>
+        Some(Point(coords.head._1, coords.head._2))
+      case coords =>
+        val line = Line(coords)
+
+        Some(isArea match {
+          case true if line.isClosed && coords.length >= 4 => Polygon(line)
+          case _ => line
+        })
+    }
+
+    geom match {
+      // drop invalid geometries
+      case Some(g) if g.isValid => g.toWKB(4326)
+      case _ => null
+    }
+  })
+
+  // create fully-formed rings from line segments
+  @tailrec
+  private def connectSegments(segments: List[Array[Coordinate]], rings: List[Array[Coordinate]] = List.empty[Array[Coordinate]]): List[Array[Coordinate]] = {
+    segments match {
+      case Nil => rings
+      case h :: t if h.head equals2D h.last => connectSegments(t, rings :+ h)
+      case h :: t =>
+        connectSegments(t.find(line => h.last equals2D  line.head) match {
+          case Some(next) => h ++ next.tail :: t.filterNot(line => line sameElements next)
+          case None =>
+            t.find(line => h.last == line.last) match {
+              case Some(next) => h ++ next.reverse.tail :: t.filterNot(line => line sameElements next)
+              case None => throw new Exception("Unable to connect segments.")
+            }
+        }, rings)
+    }
+  }
+
+  private def connectSegments(segments: List[Line]): List[Polygon] = {
+    connectSegments(segments.map(_.jtsGeom.getCoordinates): List[Array[Coordinate]]).map(_.map(Point.jtsCoord2Point)).map(Polygon(_))
```

Comment:
`jtsCoord2Point` is an implicit conversion; how do I use it implicitly?

Reply:
You could just use `segments.map(_.vertices)` instead of `segments.map(_.jtsGeom.getCoordinates)`

Reply:
I want to avoid calling `vertices` to prevent creating additional objects; it clones and JTS generates `Points` underneath, so using raw `Coordinate`s is better from a GC standpoint.

Reply:
Gotcha, mutability vs performance strikes again.

```scala
connectSegments(segments.map(_.jtsGeom.getCoordinates.map(_: Point))).map(Polygon(_))
```

is how I would approach it. The type tag kicks the implicit in.

Reply:
```
connectSegments(segments.map(_.jtsGeom.getCoordinates): List[Array[Coordinate]]).map(cs => cs.map(_: Point)).map(Polygon(_))
```

seems like it should work, but the compiler complains with:

```
Error:(201, 104) type mismatch;
 found   : geotrellis.vector.Point
 required: com.vividsolutions.jts.geom.Coordinate => ?
    connectSegments(segments.map(_.jtsGeom.getCoordinates): List[Array[Coordinate]]).map(cs => cs.map(_: Point)).map(Polygon(_))
```

Maybe what I actually need is a converter between `Seq[Coordinate]` → `Seq[Point]`?

(This doesn't actually matter, though learning about implicit conversions is helpful.)